libp2p/connmanager.nim

```diff
@@ -331,42 +355,119 @@ proc storeConn*(c: ConnManager, conn: Connection) =
   ##
 
   if isNil(conn):
-    raise newException(CatchableError, "connection cannot be nil")
+    raise newException(CatchableError, "Connection cannot be nil")
 
-  if conn.closed() or conn.atEof():
-    trace "Can't store dead connection", conn
-    raise newException(CatchableError, "can't store dead connection")
+  if conn.closed or conn.atEof:
+    raise newException(CatchableError, "Connection closed or EOF")
 
   if isNil(conn.peerInfo):
-    raise newException(CatchableError, "empty peer info")
+    raise newException(CatchableError, "Empty peer info")
 
   let peerId = conn.peerInfo.peerId
-  if c.conns.getOrDefault(peerId).len > c.maxConns:
-    debug "too many connections",
+  if c.conns.getOrDefault(peerId).len > c.maxConnsPerPeer:
+    debug "Too many connections for peer",
       conn, conns = c.conns.getOrDefault(peerId).len
 
-    raise newTooManyConnections()
+    raise newTooManyConnectionsError()
 
   if peerId notin c.conns:
     c.conns[peerId] = initHashSet[Connection]()
 
   c.conns[peerId].incl(conn)
+  libp2p_peers.set(c.conns.len.int64)
 
   # Launch on close listener
   # All the errors are handled inside `onClose()` procedure.
   asyncSpawn c.onClose(conn)
-  libp2p_peers.set(c.conns.len.int64)
 
   trace "Stored connection",
     conn, direction = $conn.dir, connections = c.conns.len
 
-proc storeOutgoing*(c: ConnManager, conn: Connection) =
-  conn.dir = Direction.Out
-  c.storeConn(conn)
+proc trackConn(c: ConnManager,
+               provider: ConnProvider,
+               dir: Direction):
+               Future[Connection] {.async.} =
+  var conn: Connection
+  try:
+    conn = await provider()
+
+    if isNil(conn):
+      return
+
+    conn.dir = dir
+    trace "Got connection", conn, dir = $dir
+
+    proc semaphoreMonitor() {.async.} =
+      try:
+        await conn.join()
+      except CatchableError as exc:
+        trace "Exception in semaphore monitor, ignoring", exc = exc.msg
+
+      if conn.dir == Direction.In:
+        c.inSema.release()
+      else:
+        c.outSema.release()
+
+    asyncSpawn semaphoreMonitor()
+  except CatchableError as exc:
```

Comment:
this is to catch exceptions in.. `provider()`? it's the only thing that potentially raises here, so it shouldn't be needed

Reply:
Since `provider` is user defined and we don't have a good way of annotating async methods with allowed exception types, this is to make sure both flows are covered.

Reply:
the point was more that neither `asyncSpawn` nor the monitor raises, so reraising the exception from `provider` makes the exception handler a no-op because `conn` will never be anything but `nil`